airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py

```diff
@@ -0,0 +1,167 @@
+            )
+
+        return job_status
+
+    def _start_job_and_validate_response(self, stream_slice: StreamSlice) -> requests.Response:
+        """
+        Starts a job and validates the response.
+
+        Args:
+            stream_slice (StreamSlice): The stream slice to be used for the job.
+
+        Returns:
+            requests.Response: The response from the job creation requester.
+
+        Raises:
+            AirbyteTracedException: If no response is received from the creation requester.
+        """
+
+        response: Optional[requests.Response] = self.creation_requester.send_request(stream_slice=stream_slice)
+        if not response:
+            raise AirbyteTracedException(
+                internal_message="Always expect a response or an exception from creation_requester",
+                failure_type=FailureType.system_error,
+            )
+
+        return response
+
+    def start(self, stream_slice: StreamSlice) -> AsyncJob:
+        """
+        Starts a job for the given stream slice.
+
+        Args:
+            stream_slice (StreamSlice): The stream slice to start the job for.
+
+        Returns:
+            AsyncJob: The asynchronous job object representing the started job.
+        """
+
+        response: requests.Response = self._start_job_and_validate_response(stream_slice)
+        job_id: str = str(uuid.uuid4())
+        self._create_job_response_by_id[job_id] = response
+
+        return AsyncJob(api_job_id=job_id, job_parameters=stream_slice)
+
+    def update_jobs_status(self, jobs: Iterable[AsyncJob]) -> None:
+        """
+        Updates the status of multiple jobs.
+
+        Args:
+            jobs (Iterable[AsyncJob]): An iterable of AsyncJob objects representing the jobs to update.
+
+        Returns:
+            None
+        """
+        for job in jobs:
+            stream_slice = StreamSlice(
+                partition={"create_job_response": self._create_job_response_by_id[job.api_job_id()]},
+                cursor_slice={},
+            )
+            polling_response: requests.Response = self._get_validated_polling_response(stream_slice)
+            job_status: AsyncJobStatus = self._get_validated_job_status(polling_response)
+            job.update_status(job_status)
+            if job_status == AsyncJobStatus.COMPLETED:
+                self._polling_job_response_by_id[job.api_job_id()] = polling_response
+
+    def fetch_records(self, job: AsyncJob) -> Iterable[Mapping[str, Any]]:
+        """
+        Fetches records from the given job.
+
+        Args:
+            job (AsyncJob): The job to fetch records from.
+
+        Yields:
+            Iterable[Mapping[str, Any]]: A generator that yields records as dictionaries.
+
+        """
+
+        for url in self.urls_extractor.extract_records(self._polling_job_response_by_id[job.api_job_id()]):
+            stream_slice: StreamSlice = StreamSlice(partition={"url": url}, cursor_slice={})
```

Comment:
Because we don't have interpolation on random fields, we have this hack which consist on using the stream_slice to allow for interpolation. This is what it looks like in the manifest: https://github.com/airbytehq/airbyte/compare/async-job/feature-branch#diff-b041c21db61fd5fe7fb67481e6167b6a343236571089c1ed61498b597202a750R921

Reply:
Q: would it make sense to create a new type of slice or partition since this isn't exactly a StreamSlice in the typical sense?

Reply:
The answer for me would be the allow for interpolation for more parameters than the stream_slice, stream_state and next_page_token as [we limit the interpolation to these](https://github.com/airbytehq/airbyte/blob/58989b04f5542ebd9d245860b4d9ca062f2a5630/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/request_options/interpolated_request_input_provider.py#L51). This is can be done but will require a change to those three layers: HttpRequester, RequestOptionProvider, RequestInputProvider